_posts/2019-12-20-differential-privacy/differential_privacy.Rmd

```diff
@@ -0,0 +1,470 @@
+s1$signal$wrist$ACC %>% nrow() # 294784
+# PPG data - sampling freq. 64 Hz
+s1$signal$wrist$BVP %>% nrow() # 589568
+# electrodermal activity data - sampling freq. 4 Hz
+s1$signal$wrist$EDA %>% nrow() # 36848
+# body temperature data - sampling freq. 4 Hz
+s1$signal$wrist$TEMP %>% nrow() # 36848
+
+### target ###
+
+# EKG data - provided in already averaged form, at frequency 0.5 Hz
+s1$label %>% nrow() # 4603
+```
+
+In light of the different sampling frequencies, our `tfdatasets` pipeline will have do some moving averaging, paralleling that applied to construct the ground truth data.
+
+
+## Preprocessing pipeline
+
+As every "column" ^[For convenience, we take the liberty to talk as if we had the usual rectangular data here.] is of different length and resolution, we build up the final _dataset_ piece-by-piece.
+The following function serves two purposes:
+
+1. compute running averages over differently sized windows, thus downsampling to 0.5Hz for every modality
+2. transform the data to the `(num_timesteps, num_features)` format that will be required by the 1d-convnet we're going to use soon
+
+```{r}
+average_and_make_sequences <-
+  function(data, window_size_avg, num_timesteps) {
+    data %>% tf$cast(tf$float32) %>%
+      # create an initial tf.data dataset to work with
+      tensor_slices_dataset() %>%
+      # use dataset_window to compute the running average of size window_size_avg
+      dataset_window(window_size_avg) %>%
+      dataset_flat_map(function (x)
+        x$batch(as.integer(window_size_avg), drop_remainder = TRUE)) %>%
+      dataset_map(function(x)
+        tf$reduce_mean(x, axis = 0L)) %>%
+      # use dataset_window to create a "timesteps" dimension with length num_timesteps)
+      dataset_window(num_timesteps, shift = 1) %>%
+      dataset_flat_map(function(x)
+        x$batch(as.integer(num_timesteps), drop_remainder = TRUE))
+  }
+```
+
+We'll call this function for every column separately. Not all columns are exactly the same length (in terms of time), thus it's safest to cut off individual observations that surpass a common length (dictated by the target variable):
+
+```{r}
+label <- s1$label %>% matrix() # 4603 observations, each spanning 2 secs
+n_total <- 4603 # keep track of this
+
+# keep matching numbers of observations of predictors
+acc <- s1$signal$wrist$ACC[1:(n_total * 64), ] # 32 Hz, 3 columns
+bvp <- s1$signal$wrist$BVP[1:(n_total * 128)] %>% matrix() # 64 Hz
+eda <- s1$signal$wrist$EDA[1:(n_total * 8)] %>% matrix() # 4 Hz
+temp <- s1$signal$wrist$TEMP[1:(n_total * 8)] %>% matrix() # 4 Hz
+```
+
+Some more housekeeping. Both training and the test set need to have a `timesteps` dimension, as usual with architectures that work on sequential data (1-d convnets and RNNs). To make sure there is no overlap between respective `timesteps`, we split the data "up front" and assemble both sets separately. We'll use the first 4000 observations for training.
```

Comment:
I think a line chart somewhere of some of these sequential predictors would be helpful in getting oriented.

Reply:
I think I'll skip this one, as probably the dataset related content is on the verge of distracting too much from the main topic already...